File: packages/sdk/js/packages/admin/src/push.ts
/**
 * PushClient — Push notification management for Admin/Server SDK
 *
 * Usage:
 *   const admin = createAdminClient(url, { serviceKey });
 *   await admin.push.send('userId', { title: 'Hello', body: 'World' });
 *   await admin.push.sendMany(['u1', 'u2'], { title: 'News' });
 *   const logs = await admin.push.getLogs('userId');
 */
import type { HttpClient } from '@edgebase/core';
import { HttpClientAdapter } from '@edgebase/core';
import { DefaultAdminApi } from './generated/admin-api-core.js';

// ─── Types ───

export interface PushPayload {
  title?: string;
  body?: string;
  image?: string;
  sound?: string;
  badge?: number;
  data?: Record<string, unknown>;
  silent?: boolean;
  collapseId?: string;
  ttl?: number;
  fcm?: Record<string, unknown>;
}

export interface PushResult {
  sent: number;
  failed: number;
  removed: number;
}

export interface PushLogEntry {
  sentAt: string;
  userId: string;
  platform: string;
  status: 'sent' | 'failed' | 'removed';
  collapseId?: string;
  error?: string;
}

/** Device token info returned by getTokens() — token value is NOT exposed. */
export interface DeviceTokenInfo {
  deviceId: string;
  platform: string;
  updatedAt: string;
  deviceInfo?: {
    name?: string;
    osVersion?: string;
    appVersion?: string;
    locale?: string;
  };
  metadata?: Record<string, unknown>;
}

// ─── PushClient ───

export class PushClient {
  private adminCore: DefaultAdminApi;

  constructor(httpClient: HttpClient) {
    this.adminCore = new DefaultAdminApi(new HttpClientAdapter(httpClient));
  }

  /** Send a push notification to a single user's devices. */
  async send(userId: string, payload: PushPayload): Promise<PushResult> {
    return this.adminCore.pushSend({ userId, payload }) as Promise<PushResult>;
  }

  /** Send a push notification to multiple users (no limit — server chunks internally). */
  async sendMany(userIds: string[], payload: PushPayload): Promise<PushResult> {
    return this.adminCore.pushSendMany({ userIds, payload }) as Promise<PushResult>;
  }

  /** Send a push notification directly to a specific FCM token. */
  async sendToToken(token: string, payload: PushPayload, platform?: string): Promise<PushResult> {
    return this.adminCore.pushSendToToken({ token, payload, platform }) as Promise<PushResult>;
  }

  /** Get registered device tokens for a user -- token values are NOT exposed. */
  async getTokens(userId: string): Promise<DeviceTokenInfo[]> {
    const query: Record<string, string> = { userId };
    const res = await this.adminCore.getPushTokens(query) as { items: DeviceTokenInfo[] };
    return res.items;
  }

  /** Get push send logs for a user (last 24 hours). */
  async getLogs(userId: string, limit?: number): Promise<PushLogEntry[]> {
    const query: Record<string, string> = { userId };
    if (limit !== undefined) query.limit = String(limit);
    const res = await this.adminCore.getPushLogs(query) as { items: PushLogEntry[] };
    return res.items;
  }

  /** Send a push notification to an FCM topic. */
  async sendToTopic(topic: string, payload: PushPayload): Promise<{ success: boolean; error?: string }> {
    return this.adminCore.pushSendToTopic({ topic, payload }) as Promise<{ success: boolean; error?: string }>;
  }

  /** Broadcast a push notification to all devices via /topics/all. */
  async broadcast(payload: PushPayload): Promise<{ success: boolean; error?: string }> {
    return this.adminCore.pushBroadcast({ payload }) as Promise<{ success: boolean; error?: string }>;
  }
}

File: packages/sdk/rust/packages/admin/src/push.rs
//! PushClient — Push notification management for Admin SDK.

use std::sync::Arc;

use serde_json::Value;

use crate::generated::admin_api_core::GeneratedAdminApi;
use edgebase_core::error::Error;
use edgebase_core::http_client::HttpClient;

/// Result of a push send operation.
#[derive(Debug, Default)]
pub struct PushResult {
    pub sent: i64,
    pub failed: i64,
    pub removed: i64,
}

/// Client for push notification operations.
pub struct PushClient {
    http: Arc<HttpClient>,
}

impl PushClient {
    pub(crate) fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    fn core(&self) -> GeneratedAdminApi<'_> {
        GeneratedAdminApi::new(&self.http)
    }

    /// Send a push notification to a single user's devices.
    pub async fn send(&self, user_id: &str, payload: &Value) -> Result<PushResult, Error> {
        let body = serde_json::json!({
            "userId": user_id,
            "payload": payload,
        });
        let res: Value = self.core().push_send(&body).await?;
        Ok(parse_push_result(&res))
    }

    /// Send a push notification to multiple users (no limit — server chunks internally).
    pub async fn send_many(&self, user_ids: &[&str], payload: &Value) -> Result<PushResult, Error> {
        let body = serde_json::json!({
            "userIds": user_ids,
            "payload": payload,
        });
        let res: Value = self.core().push_send_many(&body).await?;
        Ok(parse_push_result(&res))
    }

    /// Send a push notification to a specific device token.
    pub async fn send_to_token(&self, token: &str, payload: &Value, platform: Option<&str>) -> Result<PushResult, Error> {
        let body = serde_json::json!({
            "token": token,
            "payload": payload,
            "platform": platform.unwrap_or("web"),
        });
        let res: Value = self.core().push_send_to_token(&body).await?;
        Ok(parse_push_result(&res))
    }

    /// Get registered device tokens for a user — token values NOT exposed.
    pub async fn get_tokens(&self, user_id: &str) -> Result<Vec<Value>, Error> {
        let mut query = std::collections::HashMap::new();
        query.insert("userId".to_string(), user_id.to_string());
        let res: Value = self.core().get_push_tokens(&query).await?;
        if let Some(items) = res.get("items").and_then(|v| v.as_array()) {
            Ok(items.clone())
        } else {
            Ok(Vec::new())
        }
    }

    /// Get push send logs for a user (last 24 hours).
    pub async fn get_logs(&self, user_id: &str, limit: Option<u32>) -> Result<Vec<Value>, Error> {
        let mut query = std::collections::HashMap::new();
        query.insert("userId".to_string(), user_id.to_string());
        if let Some(l) = limit {
            query.insert("limit".to_string(), l.to_string());
        }
        let res: Value = self.core().get_push_logs(&query).await?;
        if let Some(items) = res.get("items").and_then(|v| v.as_array()) {
            Ok(items.clone())
        } else {
            Ok(Vec::new())
        }
    }

    /// Send a push notification to an FCM topic.
    pub async fn send_to_topic(&self, topic: &str, payload: &Value) -> Result<Value, Error> {
        let body = serde_json::json!({
            "topic": topic,
            "payload": payload,
        });
        self.core().push_send_to_topic(&body).await
    }

    /// Broadcast a push notification to all devices via /topics/all.
    pub async fn broadcast(&self, payload: &Value) -> Result<Value, Error> {
        let body = serde_json::json!({
            "payload": payload,
        });
        self.core().push_broadcast(&body).await
    }
}

fn parse_push_result(v: &Value) -> PushResult {
    PushResult {
        sent: v.get("sent").and_then(|v| v.as_i64()).unwrap_or(0),
        failed: v.get("failed").and_then(|v| v.as_i64()).unwrap_or(0),
        removed: v.get("removed").and_then(|v| v.as_i64()).unwrap_or(0),
    }
}
